llvm/lib/Target/RISCV/RISCVVectorMaskDAGMutation.cpp

```diff
@@ -68,11 +72,24 @@ class RISCVVectorMaskDAGMutation : public ScheduleDAGMutation {
 
   void apply(ScheduleDAGInstrs *DAG) override {
     SUnit *NearestUseV0SU = nullptr;
+    SmallVector<SUnit *, 2> DefMask;
     for (SUnit &SU : DAG->SUnits) {
       const MachineInstr *MI = SU.getInstr();
-      if (MI->findRegisterUseOperand(RISCV::V0, TRI))
+      if (isSoleUseCopyToV0(SU))
+        DefMask.push_back(&SU);
+
+      if (MI->findRegisterUseOperand(RISCV::V0, TRI)) {
```

Comment:
is it possible that an instruction uses V0 _and_ produces a result whose only use is a COPY? If that happens we will form a cycle. Should we prevent this by checking with `DAG->canAddEdge`?